zipkin-storage/elasticsearch/src/main/java/zipkin2/elasticsearch/internal/JsonSerializers.java

```diff
@@ -0,0 +1,256 @@
+          break;
+        case "annotations":
+          if (value != JsonToken.START_ARRAY) {
+            throw new IOException("Invalid span, expecting annotations array start, got: " +
+              value);
+          }
+          while (parser.nextToken() != JsonToken.END_ARRAY) {
+            Annotation a = parseAnnotation(parser);
+            result.addAnnotation(a.timestamp(), a.value());
+          }
+          break;
+        case "tags":
+          if (value != JsonToken.START_OBJECT) {
+            throw new IOException("Invalid span, expecting tags object, got: " + value);
+          }
+          while (parser.nextValue() != JsonToken.END_OBJECT) {
+            result.putTag(parser.currentName(), parser.getValueAsString());
+          }
+          break;
+        case "debug":
+          result.debug(parser.getBooleanValue());
+          break;
+        case "shared":
+          result.shared(parser.getBooleanValue());
+          break;
+        default:
+          // Skip
+      }
+    }
+
+    return result.build();
+  }
+
+  static Endpoint parseEndpoint(JsonParser parser) throws IOException {
+    if (!parser.isExpectedStartObjectToken()) {
+      throw new IOException("Not a valid JSON object, start token: " +
+        parser.currentToken());
+    }
+
+    String serviceName = null, ipv4 = null, ipv6 = null;
+    int port = 0;
+
+    while (parser.nextToken() != JsonToken.END_OBJECT) {
+      JsonToken value = parser.nextValue();
+      if (value == JsonToken.VALUE_NULL) {
+        continue;
+      }
+
+      switch (parser.currentName()) {
+        case "serviceName":
+          serviceName = parser.getText();
+          break;
+        case "ipv4":
+          ipv4 = parser.getText();
+          break;
+        case "ipv6":
+          ipv6 = parser.getText();
+          break;
+        case "port":
+          port = parser.getIntValue();
+          break;
+        default:
+          // Skip
+      }
+    }
+
+    if (serviceName == null && ipv4 == null && ipv6 == null && port == 0) return null;
+    return Endpoint.newBuilder()
+      .serviceName(serviceName)
+      .ip(ipv4)
+      .ip(ipv6)
+      .port(port)
+      .build();
+  }
+
+  static Annotation parseAnnotation(JsonParser parser) throws IOException {
+    if (!parser.isExpectedStartObjectToken()) {
+      throw new IOException("Not a valid JSON object, start token: " +
+        parser.currentToken());
+    }
+
+    long timestamp = 0;
+    String value = null;
+
+    while (parser.nextValue() != JsonToken.END_OBJECT) {
+      switch (parser.currentName()) {
+        case "timestamp":
+          timestamp = parser.getLongValue();
+          break;
+        case "value":
+          value = parser.getValueAsString();
+          break;
+        default:
+          // Skip
+      }
+    }
+
+    if (timestamp == 0 || value == null) {
+      throw new IllegalStateException("Incomplete annotation at " + parser.currentToken());
```

Comment:
I know we have been inconsistent, but usually use IllegalArgumentException on parsing as State is something that could resolve, and this won't.